File: Lesson08/RecyclerView/app/src/main/java/com/codingblocks/recyclerview/MainActivity.java
package com.codingblocks.recyclerview;

import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.StaggeredGridLayoutManager;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ListView;

import java.util.ArrayList;

public class MainActivity extends AppCompatActivity {

    ArrayList<Note> notesList = new ArrayList<>();
    EditText etTitle;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        Button btnAdd = findViewById(R.id.btnAdd);

        etTitle = findViewById(R.id.etNote);

        RecyclerView rv = findViewById(R.id.rvNotes);

//        rv.setLayoutManager(new LinearLayoutManager(this,LinearLayoutManager.HORIZONTAL,false));

//        rv.setLayoutManager(new GridLayoutManager(this, 2));

        rv.setLayoutManager(new StaggeredGridLayoutManager(2, StaggeredGridLayoutManager.VERTICAL));

        final NoteAdapter noteAdapter = new NoteAdapter(notesList, this);

        rv.setAdapter(noteAdapter);

        btnAdd.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                String title = etTitle.getText().toString();
                String time = "" + System.currentTimeMillis();
                notesList.add(new Note(title, time));
                noteAdapter.notifyDataSetChanged();
            }
        });
    }
}

File: Lesson16/NotesFirestore/app/src/main/java/app/harshit/notesfirestore/MainActivity.kt
package app.harshit.notesfirestore

import android.os.Bundle
import android.util.EventLog
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import com.google.firebase.firestore.*
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.layout_dialog.view.*

class MainActivity : AppCompatActivity() {

    private val notes = arrayListOf<Note>()
    private val noteAdapter by lazy {
        NoteAdapter(notes)
    }

    private val db = FirebaseFirestore.getInstance()

    private val dialogView by lazy {
        layoutInflater.inflate(R.layout.layout_dialog, null, false)
    }

    private val noteAlert by lazy {
        AlertDialog.Builder(this)
            .setTitle("Enter the note here")
            .setView(dialogView)
            .setPositiveButton("Save") { dialog, _ ->

                val note = Note(
                    dialogView.etTitle.text.toString(),
                    dialogView.etDescription.text.toString(),
                    System.currentTimeMillis().toString()
                )

                saveNoteToFireStore(note)
                dialog.dismiss()
            }
            .setNegativeButton("Cancel") { dialog, _ ->
                dialog.dismiss()
            }
            .create()
    }

    private fun saveNoteToFireStore(note: Note) {

//        db.collection("notes")
//            .document("dhfds").set(note)

//        val collectionRef = db.collection("notes")
//
//        val documentRef = collectionRef.document()
//
//        documentRef.set(note)

        db.collection("notes")
            .add(note)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        rvNotes.layoutManager = LinearLayoutManager(this)
        rvNotes.adapter = noteAdapter

        fabAddNote.setOnClickListener {
            noteAlert.show()
        }
    }

    override fun onStart() {
        super.onStart()

//        db.collection("notes")
//            .addSnapshotListener { querySnapshot, firebaseFirestoreException ->
//
//                querySnapshot?.documentChanges?.forEach {
//
//                    when (it.type) {
//                        DocumentChange.Type.ADDED -> {
//                            val currentNote = it.document
//                                .toObject(Note::class.java)
//                            //Add this to arraylist and notify the adapter
//                        }
//                        DocumentChange.Type.MODIFIED -> {
//                            val currentNote = it.document
//                                .toObject(Note::class.java)
//                            //Update the arraylist and notify the adapter
//                        }
//                        DocumentChange.Type.REMOVED -> {
//                            val currentNote = it.document
//                                .toObject(Note::class.java)
//                            //Remove item from arraylist and notify adapter
//                        }
//                    }
//                }
//
//            }


        db.collection("notes")
            .addSnapshotListener { querySnapshot, firebaseFirestoreException ->
                querySnapshot?.documentChanges?.forEach {
                    notes.add(0, it.document.toObject(Note::class.java))
                    noteAdapter.notifyItemInserted(0)
                    rvNotes.scrollToPosition(0)
                }
            }

//        db.collection("notes").get()
//            .addOnSuccessListener {
//                it.documents.forEach {
//                    notes.add(it.toObject(Note::class.java)!!)
//                }
//
//                noteAdapter.notifyDataSetChanged()
//            }

    }

}
